README.md

```diff
@@ -1,32 +1,224 @@
+of `Error`.
+
+Recall that curly braces are required to create a new code block in Java.
+Indentation is optional and does not create a code block, although its use
+is encouraged for readability. Thus, the body of the `while` loop in the code
+below contains only the statement `list.add(i);`
+and not `i++;` due to missing curly braces around the two statements.
+Since `i++;` is never executed, the code results in an infinite loop that keeps adding
+elements to the list until the program runs out of memory.
+
+```java
+import java.util.ArrayList;
+import java.util.List;
+
+public class OutOfMemoryErrorExample {
+
+    public static void main(String[] args) {
+        int i=0;
+        List<Integer> list = new ArrayList<>();
+        // while loop is missing curly braces to create a code block
+        while (i<10)
+            list.add(i);
+            i++; 	//this statement never reached due to missing curly braces
+        System.out.println("done");
+    }
+
+}
+```
+
+Eventually the Java Virtual Machine runs out of memory and crashes:
+
+```text
+Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
+	at java.base/java.util.Arrays.copyOf(Arrays.java:3512)
+	at java.base/java.util.Arrays.copyOf(Arrays.java:3481)
+	at java.base/java.util.ArrayList.grow(ArrayList.java:237)
+	at java.base/java.util.ArrayList.grow(ArrayList.java:244)
+	at java.base/java.util.ArrayList.add(ArrayList.java:454)
+	at java.base/java.util.ArrayList.add(ArrayList.java:467)
+	at OutOfMemoryErrorExample.main(OutOfMemoryErrorExample.java:11)
+```
+
+## `Exception`
+
+The signature of the `Scanner`  constructor indicates it throws `FileNotFoundException`:
+
+```java
+public Scanner(File source) throws FileNotFoundException
+```
+
+The `Scanner` constructor will attempt to open the file passed as a parameter,
+and an exception is thrown if the file does not exist.
+`FileNotFoundException` is a subclass of `Exception`,
+which means it is a checked exception that must be handled.
+
+The following class fails to compile because it calls the `Scanner` constructor,
+but does not attempt to catch the potential `FileNotFoundException` that might be thrown.
+
+```java
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+public class FileNotFoundExceptionExample {
+
+    public static void readFile() {
+        Scanner input = null;
+        input = new Scanner(new File("data.txt"));  //may throw FileNotFoundException
+        String text = input.next();
+        System.out.println(text);
+    }
+
+    public static void main(String[] args) {
+        readFile();
+    }
+}
+```
+
+We can fix this by placing the constructor call in a `try-catch` block:
+
+```java
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+public class FileNotFoundExceptionExample {
+
+  public static void readFile() {
+    // try with resources
+    try (Scanner input = new Scanner(new File("data.txt")))
```

Comment:
I didn't show a try-with-resources block in Java Mod 1. The first time I had shown this was Working With Files, which has now been pushed back to Java Mod 4; so this might need to be further explained unless it was explained in an earlier lesson in Mod 2 or 3 at this point.

Reply:
remove try-with-resources and just used basic try-catch.